test/testground/network/sync.go

```diff
@@ -0,0 +1,218 @@
+	CommandTopic = sync.NewTopic("command", Command{})
+)
+
+// PeerPacket is the message that is sent to other nodes upon network
+// initialization. It contains the necessary info from this node to start the
+// network.
+type PeerPacket struct {
+	PeerID          string          `json:"peer_id"`
+	GroupID         string          `json:"group_id"`
+	GlobalSequence  int64           `json:"global_sequence"`
+	GenesisAccounts []string        `json:"genesis_accounts"`
+	GenesisPubKeys  []string        `json:"pub_keys"`
+	GenTx           json.RawMessage `json:"gen_tx"`
+}
+
+func (pp *PeerPacket) IsValidator() bool {
+	return pp.GroupID == ValidatorGroupID
+}
+
+func (pp *PeerPacket) IsLeader() bool {
+	return pp.GlobalSequence == LeaderGlobalSequence
+}
+
+func (pp *PeerPacket) Name() string {
+	return NodeID(pp.GlobalSequence)
+}
+
+func (pp *PeerPacket) GetPubKeys() ([]cryptotypes.PubKey, error) {
+	pks := make([]cryptotypes.PubKey, 0, len(pp.GenesisPubKeys))
+	for _, pk := range pp.GenesisPubKeys {
+		sdkpk, err := DeserializeAccountPublicKey(pk)
+		if err != nil {
+			return nil, err
+		}
+		pks = append(pks, sdkpk)
+	}
+	return pks, nil
+}
+
+func NewConfigSet(params *Params, pps []PeerPacket) []RoleConfig {
+	nodeConfigs := make([]RoleConfig, 0, len(pps))
+
+	for _, pp := range pps {
+		nodeConfigs = append(nodeConfigs, RoleConfig{
+			CmtConfig:      StandardCometConfig(params),
+			AppConfig:      StandardAppConfig(params),
+			PeerID:         pp.PeerID,
+			GroupID:        pp.GroupID,
+			GlobalSequence: pp.GlobalSequence,
+		})
+	}
+
+	return nodeConfigs
+}
+
+type RoleConfig struct {
+	PeerID         string            `json:"peer_id"`
+	GroupID        string            `json:"group_id"`
+	GlobalSequence int64             `json:"global_sequence"`
+	CmtConfig      *tmconfig.Config  `json:"cmt_config"`
+	AppConfig      *srvconfig.Config `json:"app_config"`
+}
+
+func peerIDs(nodes []RoleConfig) []string {
+	peerIDs := make([]string, 0, len(nodes))
+	for _, nodeCfg := range nodes {
+		peerIDs = append(peerIDs, nodeCfg.PeerID)
+	}
+	return peerIDs
+}
+
+func searchNodes(nodes []RoleConfig, globalSeq int64) (RoleConfig, bool) {
+	for _, node := range nodes {
+		if node.GlobalSequence == globalSeq {
+			return node, true
+		}
+	}
+	return RoleConfig{}, false
+}
+
+func PublishNodeConfigs(ctx context.Context, initCtx *run.InitContext, nodes []RoleConfig) error {
+	for _, node := range nodes {
+		_, err := initCtx.SyncClient.Publish(ctx, MetaConfigTopic, node)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func DownloadNodeConfigs(ctx context.Context, runenv *runtime.RunEnv, initCtx *run.InitContext) ([]RoleConfig, error) {
+	return DownloadSync(ctx, initCtx, MetaConfigTopic, RoleConfig{}, runenv.TestInstanceCount)
+}
+
+// GenesisWrapper is a simple struct wrapper that makes it easier testground to
+// properly serialize and distribute the genesis file.
+type GenesisWrapper struct {
+	Part    int    `json:"part"`
```

Comment:
Would be great to add a comment about what the `Part` field is for.

Reply:
261214a5adbc10044fac9c6c3b4b07ba18c2aea0